### DISCUSSION TITLE
New WFDB header format - layout and metadata

### DISCUSSION THREAD

**Discussion by cx1111** [wfdb/wfdb-spec#14] [General]
## 1. Header File

Currently, WFDB header files use the `.hea` file extension, and have their own [layout](https://github.com/wfdb/wfdb-spec/blob/main/spec/HEADERS.md). The fields should be human-readable, but in reality, it is hard for someone who hasn't worked extensively with WFDB to parse them. eg. <https://physionet.org/content/mitdb/1.0.0/100.hea> 

There are benefits of having a standalone WFDB header file, as listed in #12. Therefore it may be worth creating a new header format that keeps these benefits, while being easier to use by new and experienced users alike. We could utilize existing widely-used file formats which show key-value pairs. eg.
- YAML
- TOML
- (Probably not JSON or XML.)

This could be a complementary change to the single file change in discussion #12. We could have a new header format, and wrap it up in a wrapper file. Note: if having a single `.wfdb` file solves a bunch of clarity issues, while forcing the metadata to be read by WFDB applications instead of text editors/browsers, creating a new clearer header file will be of less value. At this point however, I don't know which, if either, will be implemented, so I think it's worth considering both items.

## 2. Header Fields

The second issue regards the structured metadata that should be present in the header. 

From an earlier discussion, it was generally agreed that patient metadata, such as age, sex, weight, etc. don't belong in these files used for storing signal data. The current WFDB software package treats certain [header comment lines](https://github.com/wfdb/wfdb-spec/blob/main/spec/HEADERS.md#comment-lines) as "info strings" and some applications such as the PhysioBank signal search tool, will explicitly try to parse these fields. This seems like a bad practice which relies on an implicit convention, that is never enforced. Although we currently have no alternative for searching through signals and filtering by patient metadata on PhysioNet, we should definitely be able to come up with a better solution/convention, that doesn't rely on the header files. eg. Require a `patient-metadata.csv` file for each dataset.

Despite there being precedence from some formats such as DICOM,  including such fields are likely to create more needless complication than benefits. ie. A needlessly complex format will be harder to de-identify, and as part of a database, will form a conflicting source of ground-truth for fields.

To summarize, I do NOT support having structured fields for general patient metadata.

From a database management perspective however, it may be useful to include structured fields for:
- Person/Patient ID
- Device ID

**Comment by cx1111** [comment]
As a general comment, the current header contains many optional fields that can only be present if another field is present. Sometimes this may make sense, but in other times, it seems to be only for backwards compatibility reasons? eg. The "Description" (signal name) field can only be present if the "Block size" field is present, yet they don't appear to be functionally related at all.

This is rather confusing behavior which we would not want to replicate in a future header format.

**Comment by tompollard** [comment]
I'm going to try to make some progress on this, and will mostly be posting at: https://github.com/wfdb/wfdb-spec/issues/22
